Add JSON-RPC access to directories and server lists (without connect/disconnect methods) (#3479)
jamulussoftware/jamulus: docs/JSON-RPC.md

> | result.version | string | The Jamulus version. |

Reviewer: Consistency on a page is a mercy for readers, and this page consistently puts a period behind non-sentences. Ideally, none would include a period.

> ### jamulus/pollServerList

Author: ```suggestion
### jamulusclient/pollServerList
```

> Request list of servers in a directory

Reviewer: ```suggestion
Request list of servers in a directory.
```

Parameters:

| Name | Type | Description |
| --- | --- | --- |

> | params.directory | string | socket address of directory to query, e.g. anygenre1.jamulus.io:22124. |

Reviewer: ```suggestion
| params.directory | string | socket address of directory to query. Example: anygenre1.jamulus.io:22124. |
```

e.g. is not used elsewhere on this page. Perhaps it could be avoided here.

| result.id | number | The channel ID. |
| result.name | string | The musician’s name. |
| result.skillLevel | string | The musician’s skill level (beginner, intermediate, expert, or null). |

> | result.country | string | The musician’s country. |

Author: https://github.com/jamulussoftware/jamulus/pull/3249#issuecomment-2030271992

I'd suggest to add the id back...

Reviewer 2: Definitely. No problem with adding fields to the JSON, which can be used or ignored, but we should not remove existing fields that a JSON-RPC client might potentially be using already. So retain `countryId` and `instrumentId`, and add `country` and `instrument` as new fields.

Reviewer 2: ```suggestion
| result.countryId | number | The musician’s country ID (see QLocale::Country). |
| result.country | string | The musician’s country. |
```

| result.city | string | The musician’s city. |

> | result.instrument | number | The musician’s instrument. |

Reviewer 2: ```suggestion
| result.instrumentId | number | The musician’s instrument ID (see CInstPictures::GetTable). |
| result.instrument | string | The musician’s instrument. |
```

| params.clients[*].id | number | The channel ID. |
| params.clients[*].name | string | The musician’s name. |
| params.clients[*].skillLevel | string | The musician’s skill level (beginner, intermediate, expert, or null). |

> | params.clients[*].country | string | The musician’s country. |

Author: As above.

Reviewer 2: ```suggestion
| params.clients[*].countryId | number | The musician’s country ID (see QLocale::Country). |
| params.clients[*].country | string | The musician’s country. |
```

| params.clients[*].city | string | The musician’s city. |

> | params.clients[*].instrument | string | The musician’s instrument. |

Reviewer 2: ```suggestion
| params.clients[*].instrumentId | number | The musician’s instrument ID (see CInstPictures::GetTable). |
| params.clients[*].instrument | string | The musician’s instrument. |
```

| params | object | No parameters (empty object). |

### jamulusclient/recorderState

> Emitted when the client is connected to a server who's recorder state changes.

Reviewer 2: ```suggestion
Emitted when the client is connected to a server whose recorder state changes.
```

Parameters:

| Name | Type | Description |
| --- | --- | --- |
| params.state | number | The recorder state |

### jamulusclient/serverInfoReceived

Emitted when a server info is received.

Parameters:

| Name | Type | Description |
| --- | --- | --- |
| params.address | string | The server socket address |
| params.pingtime | number | The round-trip ping time in ms |
| params.numClients | number | The quantity of clients connected to the server |

### jamulusclient/serverListReceived

Emitted when the server list is received.

Parameters:

| Name | Type | Description |
| --- | --- | --- |
| params.servers | array | The server list. |
| params.servers[*].address | string | Socket address (ip_address:port) |
| params.servers[*].name | string | Server name |

> | params.servers[*].country | string | Server country |

Reviewer 2: Also include `countryId` for consistency with `jamulusclient/clientListReceived` suggestion above.

Reviewer 2: ```suggestion
| params.servers[*].countryId | number | Server country ID (see QLocale::Country). |
| params.servers[*].country | string | Server country |
```

| params | object | No parameters (empty object). |

### jamulusclient/recorderState

Emitted when the client is connected to a server whose recorder state changes.

Parameters:

| Name | Type | Description |
| --- | --- | --- |

> | params.state | number | The recorder state |

Reviewer: ```suggestion
| params.state | number | The recorder state. |
```

I hate this fix, but it's consistent with the rest of the page.

### jamulusclient/serverInfoReceived

Emitted when a server info is received.

Parameters:

| Name | Type | Description |
| --- | --- | --- |

> | params.address | string | The server socket address |

Reviewer: ```suggestion
| params.address | string | The server socket address. |
```

> | params.pingtime | number | The round-trip ping time in ms |

Reviewer: ```suggestion
| params.pingtime | number | The round-trip ping time in ms. |
```

> | params.numClients | number | The quantity of clients connected to the server |

Reviewer: ```suggestion
| params.numClients | number | The number of clients connected to the server. |
```

Perhaps there's a notion that the definition of something shouldn't use the same word. Maybe. But the only existing similar usage on this page is result.clients[*].channels, and it says number rather than quantity.

Reviewer 3: (And the aim is to contextualise "number of clients" rather than explain "number", "of" or "clients".)

Reviewer 4: Sorry! I tend to use the word, "quantity" in docs to avoid ambiguity. Feel free to use the language with which you feel most comfortable.

### jamulusclient/serverListReceived

Emitted when the server list is received.

Parameters:

| Name | Type | Description |
| --- | --- | --- |
| params.servers | array | The server list. |

> | params.servers[*].address | string | Socket address (ip_address:port) |

Reviewer: ```suggestion
| params.servers[*].address | string | Socket address (ip_address:port). |
```

> | params.servers[*].name | string | Server name |

Reviewer: ```suggestion
| params.servers[*].name | string | Server name. |
```

| params.servers[*].countryId | number | Server country ID (see QLocale::Country). |

> | params.servers[*].country | string | Server country |

Reviewer: ```suggestion
| params.servers[*].country | string | Server country. |
```
